File: plugins/gooddata-to-sigma/skills/gooddata-to-sigma/tests/test_workbook_code_release.py
#!/usr/bin/env python3
"""Focused offline regression for the Aug-2026 workbook code release."""
import copy
import json
import os
import re
import subprocess
import sys
import tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
SKILL = os.path.dirname(HERE)
BUILDER = os.path.join(SKILL, "scripts", "build_workbook.py")
CONVERTER = os.path.join(SKILL, "scripts", "convert.py")
FIXTURE = os.path.join(SKILL, "fixtures", "test_workspace_orders.json")
sys.path.insert(0, os.path.join(SKILL, "scripts", "lib"))
import code_rep  # noqa: E402


def measure(mid, title):
    return {
        "measure": {
            "localIdentifier": mid,
            "title": title,
            "definition": {
                "measureDefinition": {
                    "item": {"identifier": {"id": mid, "type": "metric"}}
                }
            },
        }
    }


def attribute(aid):
    return {
        "attribute": {
            "localIdentifier": aid,
            "displayForm": {
                "identifier": {"id": aid + ".name", "type": "label"}
            },
        }
    }


def widget(iid, width=6):
    return {
        "size": {"xl": {"gridWidth": width}},
        "widget": {
            "insight": {"identifier": {"id": iid, "type": "visualizationObject"}}
        },
    }


def release_workspace():
    ws = copy.deepcopy(json.load(open(FIXTURE)))
    ws["analytics"]["visualizationObjects"] = [
        {
            "id": "wf", "title": "Revenue Change",
            "content": {
                "visualizationUrl": "local:waterfall",
                "properties": {
                    "controls": {
                        "legend": {"enabled": True, "position": "bottom"},
                        "style": {"backgroundColor": "#F8FAFC"},
                    }
                },
                "buckets": [
                    {"localIdentifier": "measures",
                     "items": [measure("m_net_revenue", "Net Revenue")]},
                    {"localIdentifier": "view",
                     "items": [attribute("order_channel")]},
                ],
            },
        },
        {
            "id": "rep", "title": "Revenue by Acquisition",
            "content": {
                "visualizationUrl": "local:repeater",
                "buckets": [
                    {"localIdentifier": "rows",
                     "items": [attribute("acquisition_channel")]},
                    {"localIdentifier": "columns",
                     "items": [measure("m_net_revenue", "Net Revenue")]},
                ],
            },
        },
        {
            "id": "box", "title": "Gated Distribution",
            "content": {
                "visualizationUrl": "local:boxplot",
                "buckets": [
                    {"localIdentifier": "measures",
                     "items": [measure("m_net_revenue", "Net Revenue")]},
                    {"localIdentifier": "view",
                     "items": [attribute("order_channel")]},
                ],
            },
        },
    ]
    ws["analytics"]["attributeHierarchies"] = [{
        "id": "sales_path", "title": "Sales Path",
        "attributes": ["acquisition_channel", "order_channel"],
    }]
    ws["analytics"]["analyticalDashboards"] = [{
        "id": "release_dashboard", "title": "Release Dashboard",
        "content": {
            "tabs": [
                {"localIdentifier": "overview", "title": "Overview",
                 "sections": [{"items": [widget("wf", 12)]}]},
                {"localIdentifier": "detail", "title": "Detail",
                 "sections": [{"items": [widget("rep", 12)]}]},
            ]
        },
    }]
    return ws


def test_released_workbook_shape_and_features():
    with tempfile.TemporaryDirectory() as d:
        workspace = os.path.join(d, "workspace.json")
        output = os.path.join(d, "workbook.json")
        gaps = os.path.join(d, "feature-gaps.json")
        json.dump(release_workspace(), open(workspace, "w"))
        run = subprocess.run(
            [sys.executable, BUILDER, "--workspace", workspace,
             "--data-model-id", "dm-x", "--fact-element", "el-x",
             "--fact-name", "ORDER_FACT", "--rel-name", "EL_CUSTOMER",
             "--fact-dataset", "order", "--folder-id", "folder-x",
             "--feature-gaps-out", gaps, "--out", output],
            capture_output=True, text=True,
        )
        assert run.returncode == 0, run.stdout + run.stderr
        envelope = json.load(open(output))
        gap_ledger = json.load(open(gaps))

    assert set(envelope) == {"name", "folderId", "document"}
    doc = code_rep.document(envelope)
    assert doc["kind"] == "workbook" and doc["schemaVersion"] == 1
    assert doc["layout"] and all("elements" not in page for page in doc["pages"])
    assert "<Element " in doc["layout"]
    assert not re.search(r"</?(?:LayoutElement|GridContainer)\b", doc["layout"])
    assert [page["name"] for page in doc["pages"]] == [
        "Data", "Overview", "Detail"]
    assert doc["panels"] == [] and doc["overlays"] == []
    assert doc["settings"]["navigation"]["pageTabsInViewMode"] == "shown"

    elements = code_rep.workbook_elements(envelope)
    ids = [element["id"] for element in elements]
    placed = [
        element_id
        for page_ids in code_rep.workbook_page_element_ids(envelope).values()
        for element_id in page_ids
    ]
    assert sorted(ids) == sorted(placed)
    assert len(ids) == len(set(ids)) == len(placed)

    waterfall = next(element for element in elements if element.get("id") == "wf")
    assert waterfall["kind"] == "waterfall-chart"
    assert waterfall["waterfallShape"] == {
        "calculation": "sum", "connectorLine": "shown"}
    assert waterfall["startPoint"]["value"] == {"type": "constant", "value": 0}
    assert waterfall["legend"] == {
        "visibility": "shown", "position": "bottom"}
    assert waterfall["style"] == {"backgroundColor": "#F8FAFC"}

    repeater = next(element for element in elements if element.get("id") == "rep")
    assert repeater["kind"] == "repeated-container"
    assert repeater["source"]["kind"] == "table"
    assert "<Container " in doc["layout"]
    assert any("repeated container/Net Revenue" in element.get("body", "")
               for element in elements)
    assert len([element for element in elements
                if element.get("kind") == "navigation"]) == 2
    assert not any(element.get("kind") in ("box-chart", "box-plot")
                   for element in elements)
    assert any(gap.get("feature") == "drill"
               for gap in gap_ledger["gaps"])
    assert any(gap.get("url") == "local:boxplot"
               for gap in gap_ledger["gaps"])


def test_data_model_keeps_page_nesting():
    with tempfile.TemporaryDirectory() as d:
        output = os.path.join(d, "dm.json")
        flags = os.path.join(d, "flags.json")
        run = subprocess.run(
            [sys.executable, CONVERTER, "--workspace", FIXTURE,
             "--connection-id", "conn-x", "--db", "DEMO_DB",
             "--schema", "PUBLIC", "--folder-id", "folder-x",
             "--out", output, "--flags", flags],
            capture_output=True, text=True,
        )
        assert run.returncode == 0, run.stdout + run.stderr
        dm = json.load(open(output))
    assert "document" not in dm
    assert dm["pages"][0]["elements"]
    assert "elements" not in dm


if __name__ == "__main__":
    test_released_workbook_shape_and_features()
    test_data_model_keeps_page_nesting()
    print("ALL PASS")

File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/test-workbook-code-release.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Focused offline regression for the Aug-2026 workbook-as-code release.
require 'json'
require 'open3'
require 'tmpdir'
require 'rbconfig'
require_relative 'lib/workbook_code'
require_relative 'mechanical-specs'

RUBY = RbConfig.ruby
STORIES = File.join(__dir__, 'build-story-pages.rb')
PARSER = File.join(__dir__, 'parse-twb-layout.rb')
ANCHOR_READONLY_KEYS = %w[
  workbookId url ownerId createdBy updatedBy createdAt updatedAt latestDocumentVersion
].freeze
$failures = 0

def check(label, value)
  puts "#{value ? '  ok  ' : 'FAIL  '}#{label}"
  $failures += 1 unless value
end

# Mirror the shared verify-anchors pivot-totals PUT boundary here rather than
# forking that vendored script. This Tableau-owned release check proves the
# shared CodeRep adapter retains every currently released document collection.
def anchor_put_body(spec)
  metadata = Sigma::CodeRep.metadata(spec).reject { |key, _| ANCHOR_READONLY_KEYS.include?(key) }
  Sigma::CodeRep.wrap(Sigma::CodeRep.document(spec), extra: metadata)
end

master_columns = [
  { 'id' => 'm-region', 'name' => 'Region', 'formula' => '[Fact/Region]' },
  { 'id' => 'm-sales', 'name' => 'Sales', 'formula' => '[Fact/Sales]' }
]
release_elements = [
  {
    'id' => 'wf', 'kind' => 'waterfall-chart', 'name' => 'Change by Region',
    'source' => { 'kind' => 'table', 'elementId' => 'master' },
    'columns' => [
      { 'id' => 'wf-x', 'name' => 'Region', 'formula' => '[Master/Region]' },
      { 'id' => 'wf-y', 'name' => 'Sales', 'formula' => 'Sum([Master/Sales])' }
    ],
    'xAxis' => { 'column' => 'wf-x' }, 'yAxis' => [{ 'id' => 'wf-y' }],
    'waterfallShape' => { 'calculation' => 'sum', 'connectorLine' => 'shown' },
    'startPoint' => { 'value' => { 'type' => 'constant', 'value' => 0 }, 'visibility' => 'hidden' },
    'grouping' => 'stacked', 'legend' => { 'position' => 'right' }
  },
  {
    'id' => 'progress', 'kind' => 'progress', 'name' => 'Goal',
    'mode' => 'value', 'shape' => 'bar', 'value' => 'Sum([Master/Sales])'
  },
  { 'id' => 'nav', 'kind' => 'navigation', 'mode' => 'auto' },
  { 'id' => 'break', 'kind' => 'page-break' }
]

spec = MechanicalSpecs.build_wb_spec(
  name: 'Release Workbook', dm_id: 'dm-1', fact_eid: 'fact-1',
  master_columns: master_columns, chart_elements: release_elements,
  folder_id: 'folder-1'
)
doc = Sigma::CodeRep.document(spec)

check('outer envelope keeps create metadata', spec['name'] == 'Release Workbook' &&
      spec['folderId'] == 'folder-1' && !spec.key?('pages'))
check('document declares workbook schema', doc['schemaVersion'] == 1 && doc['kind'] == 'workbook')
check('pages are metadata-only', Array(doc['pages']).all? { |page| !page.key?('elements') })
check('elements are document-global', Array(doc['elements']).map { |element| element['id'] }.sort ==
      %w[break master nav progress wf])
check('layout is required', doc['layout'].to_s.include?('<Page'))
check('layout uses live canonical Element/Container tags',
      doc['layout'].to_s.include?('<Element') &&
      !doc['layout'].to_s.match?(%r{</?(?:LayoutElement|GridContainer)\b}))
check('canonical workbook validates', WorkbookCode.validate(spec).empty?)

anchor_spec = JSON.parse(JSON.generate(spec))
anchor_spec.merge!(
  'workbookId' => 'wb-release', 'latestDocumentVersion' => 7,
  'url' => 'https://example.invalid/workbook/wb-release'
)
anchor_doc = Sigma::CodeRep.document(anchor_spec)
anchor_doc['settings'] = { 'theme' => { 'name' => 'Release Theme' } }
anchor_doc['panels'] = [{ 'id' => 'panel-release' }]
anchor_doc['overlays'] = [{ 'id' => 'overlay-release' }]
anchor_doc['agents'] = [{ 'id' => 'agent-release' }]
anchor_flat = Sigma::CodeRep.metadata(anchor_spec).merge(anchor_doc)
anchor_element = Sigma::CodeRep.workbook_elements(anchor_flat).find { |element| element['id'] == 'wf' }
anchor_element['totals'] = { 'showGrandTotals' => false, 'showSubtotals' => false }
captured_totals = anchor_element.delete('totals')
stripped_anchor_body = anchor_put_body(anchor_flat)
anchor_element['totals'] = captured_totals
restored_anchor_body = anchor_put_body(anchor_flat)

check('shared anchor PUT boundary strips read-only response metadata',
      ANCHOR_READONLY_KEYS.none? { |key| restored_anchor_body.key?(key) })
check('shared anchor totals bracket emits valid workbook code',
      WorkbookCode.validate(stripped_anchor_body).empty? &&
      WorkbookCode.validate(restored_anchor_body).empty?)
check('shared anchor totals bracket preserves released document collections',
      %w[settings panels overlays agents].all? do |key|
        Sigma::CodeRep.document(restored_anchor_body)[key] == anchor_doc[key]
      end)
check('shared anchor totals bracket restores the complete totals value',
      Sigma::CodeRep.workbook_elements(restored_anchor_body)
        .find { |element| element['id'] == 'wf' }['totals'] == captured_totals)

ids = Sigma::CodeRep.workbook_elements(doc).filter_map { |element| element['id'] }
placed = doc['layout'].scan(/\belementId="([^"]+)"/).flatten
check('layout places every element exactly once',
      ids.sort == placed.sort && placed.length == placed.uniq.length)
page_ids = Sigma::CodeRep.workbook_page_element_ids(doc)
check('CodeRep recovers page ownership from layout',
      page_ids['page-data'] == ['master'] &&
      page_ids['page-dash'] == %w[wf progress nav break])

legacy = WorkbookCode.legacy_view(spec)
legacy['pages'].last['elements'] << { 'id' => 'repeat', 'kind' => 'repeated-container' }
legacy['panels'] = [{ 'id' => 'panel-1' }]
roundtrip = WorkbookCode.canonicalize(legacy)
roundtrip_doc = Sigma::CodeRep.document(roundtrip)
check('legacy transforms re-canonicalize without nesting elements',
      roundtrip_doc['pages'].all? { |page| !page.key?('elements') } &&
      Sigma::CodeRep.workbook_elements(roundtrip_doc).any? { |element| element['id'] == 'repeat' })
check('unknown/current document collections survive compatibility transforms',
      roundtrip_doc['panels'] == [{ 'id' => 'panel-1' }])
check('new compatibility-view elements receive authoritative placement',
      Sigma::CodeRep.workbook_page_element_ids(roundtrip_doc)['page-dash'].include?('repeat') &&
      WorkbookCode.validate(roundtrip).empty?)

# Data-model code representation is deliberately outside WorkbookCode.
data_model = {
  'schemaVersion' => 1,
  'pages' => [{ 'id' => 'dm-page', 'elements' => [{ 'id' => 'dm-table' }] }]
}
check('data-model nested pages remain unchanged',
      data_model['pages'].first['elements'].first['id'] == 'dm-table' &&
      !data_model.key?('document'))

Dir.mktmpdir('tableau-story-release') do |dir|
  input = File.join(dir, 'workbook.json')
  plan = File.join(dir, 'story-plan.json')
  output = File.join(dir, 'story-workbook.json')
  File.write(input, JSON.generate(spec))
  File.write(plan, JSON.generate([
    {
      'story' => 'Executive Story',
      'points' => [
        { 'caption' => 'Opening', 'captured_sheet' => 'Release Workbook', 'sheet_kind' => 'dashboard' }
      ]
    }
  ]))
  _stdout, stderr, status = Open3.capture3(
    RUBY, STORIES, '--story-plan', plan, '--spec', input, '--out', output
  )
  check('story-page builder accepts canonical workbook input',
        status.success? || (warn(stderr) && false))
  if status.success?
    story_spec = JSON.parse(File.read(output))
    story_doc = Sigma::CodeRep.document(story_spec)
    nav = Sigma::CodeRep.workbook_elements(story_doc).find do |element|
      element['kind'] == 'navigation' && element['id'] == 'sp1-story-navigation'
    end
    check('story-page builder emits canonical flat output',
          story_doc['pages'].all? { |page| !page.key?('elements') } &&
          WorkbookCode.validate(story_spec).empty?)
    check('story points use native page navigation',
          nav && nav['mode'] == 'manual' &&
          nav.dig('options', 0, 'destination', 'pageId') == 'page-story-1')
    check('story navigation ownership comes from layout',
          Sigma::CodeRep.workbook_page_by_element(story_doc)
            .dig('sp1-story-navigation', 'id') == 'page-story-1')
  end
end

Dir.mktmpdir('tableau-waterfall-release') do |dir|
  fixture = File.join(__dir__, 'test-fixtures', 'ground-truth.twb')
  twb = File.read(fixture)
            .sub("<mark class='Line' />", "<mark class='GanttBar' />")
            .sub("<mark class='Bar' />", "<mark class='GanttBar' />")
  input = File.join(dir, 'waterfall.twb')
  output = File.join(dir, 'layout.json')
  File.write(input, twb)
  _stdout, stderr, status = Open3.capture3(RUBY, PARSER, input, output)
  check('Tableau parser accepts waterfall fixture',
        status.success? || (warn(stderr) && false))
  if status.success?
    dashboards = JSON.parse(File.read(output))
    zones = Array(dashboards.first && dashboards.first['zones'])
    waterfall = zones.find { |zone| zone['caption'] == 'Running Sales Trend' }
    gantt = zones.find { |zone| zone['caption'] == 'Sales by Region' }
    check('GanttBar plus RUNNING_SUM maps to waterfall',
          waterfall && waterfall['chart_kind'] == 'waterfall')
    check('ordinary GanttBar remains an explicit non-waterfall gap',
          gantt && gantt['chart_kind'] == 'other')
  end
end

feature_source = File.read(File.join(__dir__, 'build-charts-from-signals.rb'))
gap_source = File.read(File.join(__dir__, 'scan-workbook-gaps.rb'))
check('released waterfall kind is mapped', feature_source.include?("'waterfall'     => 'waterfall-chart'"))
check('box-chart remains a loud manual gap',
      gap_source.include?('never emit kind:box-chart') &&
      !feature_source.match?(/['"]box(?:-chart)?['"]\s*=>\s*['"]box-chart['"]/))

puts($failures.zero? ? "\nall Tableau workbook-code release tests passed" :
                       "\n#{$failures} FAILED")
exit($failures.zero? ? 0 : 1)
